mlflow/projects/__init__.py

```diff
@@ -1,156 +1,88 @@
-        if dest_path != user_param_value:
-            data.download_uri(uri=user_param_value, output_path=dest_path)
-        return os.path.abspath(dest_path)
-
-    def compute_value(self, user_param_value, storage_dir):
-        if self.type != "path" and self.type != "uri":
-            return user_param_value
-        if self.type == "uri":
-            return self._compute_uri_value(user_param_value)
-        return self._compute_path_value(user_param_value, storage_dir)
-
-
-def _sanitize_param_dict(param_dict):
-    return {str(key): shlex_quote(str(value)) for key, value in param_dict.items()}
+def run(uri, entry_point="main", version=None, parameters=None, experiment_id=None,
+        mode=None, cluster_spec=None, git_username=None, git_password=None, use_conda=True,
+        use_temp_cwd=False, storage_dir=None, block=True):
+    """
+    Run an MLflow project from the given URI.
+
+    Supports downloading projects from Git URIs with a specified version, or copying them from
+    the file system. For Git-based projects, a commit can be specified as the `version`.
+
+    :param entry_point: Entry point to run within the project. If no entry point with the specified
+                        name is found, attempts to run the project file `entry_point` as a script,
+                        using "python" to run .py files and the default shell (specified by
+                        environment variable $SHELL) to run .sh files.
+    :param experiment_id: ID of experiment under which to launch the run.
+    :param mode: Execution mode for the run. Can be set to "local" or "databricks".
+    :param cluster_spec: Path to JSON file describing the cluster to use when launching a run on
+                         Databricks.
+    :param git_username: Username for HTTP(S) authentication with Git.
+    :param git_password: Password for HTTP(S) authentication with Git.
+    :param use_conda: If True (the default), creates a new Conda environment for the run and
+                      installs project dependencies within that environment. Otherwise, runs the
+                      project in the current environment without installing any project
+                      dependencies.
+    :param use_temp_cwd: Only used if `mode` is "local" and `uri` is a local directory.
+                         If True, copies project to a temporary working directory before running it.
+                         Otherwise (the default), runs project using `uri` (the project's path) as
+                         the working directory.
+    :param storage_dir: Only used if `mode` is local. MLflow will download artifacts from
+                        distributed URIs passed to parameters of type 'path' to subdirectories of
+                        storage_dir.
+    :param block: Whether or not to block while waiting for a run to complete. Defaults to True.
+    :return: UUID of the run
+    """
+    # TODO: Allow timeout for synchronous runs?
```

Comment:
If a user wants to timeout their main thread of execution, they should just use the asynchronous API. I don't think we should implement a timeout explicitly.